internal/xds/translator/cluster.go

```diff
@@ -314,44 +317,51 @@ func buildXdsClusterLoadAssignment(clusterName string, destSettings []*ir.Destin
 	return &endpointv3.ClusterLoadAssignment{ClusterName: clusterName, Endpoints: localities}
 }
 
-func buildTypedExtensionProtocolOptions(http2, http1Trailers bool) map[string]*anypb.Any {
+func buildTypedExtensionProtocolOptions(http2 bool, http1Opts ir.HTTP1Settings) map[string]*anypb.Any {
+	if !http2 && !http1Opts.EnableTrailers && !http1Opts.PreserveHeaderCase {
+		return nil
+	}
 	var anyProtocolOptions *anypb.Any
 
+	protocolOptions := httpv3.HttpProtocolOptions{}
 	if http2 {
-		protocolOptions := httpv3.HttpProtocolOptions{
-			UpstreamProtocolOptions: &httpv3.HttpProtocolOptions_ExplicitHttpConfig_{
-				ExplicitHttpConfig: &httpv3.HttpProtocolOptions_ExplicitHttpConfig{
-					ProtocolConfig: &httpv3.HttpProtocolOptions_ExplicitHttpConfig_Http2ProtocolOptions{},
-				},
+		protocolOptions.UpstreamProtocolOptions = &httpv3.HttpProtocolOptions_ExplicitHttpConfig_{
+			ExplicitHttpConfig: &httpv3.HttpProtocolOptions_ExplicitHttpConfig{
+				ProtocolConfig: &httpv3.HttpProtocolOptions_ExplicitHttpConfig_Http2ProtocolOptions{},
 			},
 		}
-
-		anyProtocolOptions, _ = anypb.New(&protocolOptions)
-	} else if http1Trailers {
-		// TODO: If the cluster is TLS enabled, use AutoHTTPConfig instead of ExplicitHttpConfig
-		// so that when ALPN is supported enabling trailers doesn't force HTTP/1.1
-		protocolOptions := httpv3.HttpProtocolOptions{
-			UpstreamProtocolOptions: &httpv3.HttpProtocolOptions_ExplicitHttpConfig_{
-				ExplicitHttpConfig: &httpv3.HttpProtocolOptions_ExplicitHttpConfig{
-					ProtocolConfig: &httpv3.HttpProtocolOptions_ExplicitHttpConfig_HttpProtocolOptions{
-						HttpProtocolOptions: &corev3.Http1ProtocolOptions{
-							EnableTrailers: http1Trailers,
-						},
+	} else if http1Opts.EnableTrailers || http1Opts.PreserveHeaderCase {
```

Comment:
shouldnt this be `else if http1Opts.EnableTrailers` ?

Reply:
No, the check for `PreserveHeaderCase` is needed.

Both these features need to be set in the `Http1ProtocolOptions` structure, so it needs to be filled if either of these features is enabled.